File: assets/codes/1010/game_bit_operation.py
import random
import math
import time
# import numpy as np

class Game1010:
    def __init__(self):
        self.size = 10
        self.numActions = 100
        self.tilesMask = [
            0b1,                        # 0
            0b11,                       # 1
            0b10000000001,              # 2
            0b111,                      # 3
            0b100000000010000000001,    # 4
            0b010000000011,             # 5
            0b100000000011,             # 6
            0b110000000010,             # 7
            0b110000000001,             # 8
            0b1000000000100000000010000000001, # 9
            0b1111,                     # 10
            0b110000000011,             # 11
            0b10000000001000000000100000000010000000001, # 12
            0b11111,                    # 13
            0b11100000000010000000001,  # 14
            0b11100000001000000000100,    # 15
            0b10000000001000000000111,  # 16
            0b00100000000010000000111,  # 17
            0b11100000001110000000111,  # 18
        ]
        self.tilesSize   = [1, 2, 2, 3, 3, 3, 3, 3, 3, 4, 4, 4, 5, 5, 5, 5, 5, 5, 9]
        self.tilesWidth  = [1, 2, 1, 3, 1, 2, 2, 2, 2, 1, 4, 2, 1, 5, 3, 3, 3, 3, 3]
        self.tilesHeight = [1, 1, 2, 1, 3, 2, 2, 2, 2, 4, 1, 2, 5, 1, 3, 3, 3, 3, 3]
        self.tiles = [ # for render purpose only
            [[1]], # 1 tile                 # 0
            [[1,1]], # 2 tiles              # 1
            [[1],[1]],                      # 2
            [[1,1,1]], # 3 tiles            # 3
            [[1],[1],[1]],                  # 4
            [[0,1],[1,1]],                  # 5
            [[1,0],[1,1]],                  # 6
            [[1,1],[1,0]],                  # 7
            [[1,1],[0,1]],                  # 8
            [[1],[1],[1],[1]], # 4 tiles    # 9
            [[1,1,1,1]],                    # 10
            [[1,1],[1,1]],                  # 11
            [[1],[1],[1],[1],[1]], # 5 tiles# 12
            [[1,1,1,1,1]],                  # 13
            [[1,1,1],[0,0,1],[0,0,1]],      # 14
            [[1,1,1],[1,0,0],[1,0,0]],      # 15
            [[1,0,0],[1,0,0],[1,1,1]],      # 16
            [[0,0,1],[0,0,1],[1,1,1]],      # 17
            [[1,1,1],[1,1,1],[1,1,1]]       # 18
        ]
        self.numTiles = len(self.tilesHeight)
        
        # rowMask = 0b1111111111 << i*10
        # colMask = 0b1000000000100000000010000000001000000000100000000010000000001000000000100000000010000000001 << i
        self.rowMasks = [1023, 1047552, 1072693248, 1098437885952, 1124800395214848, 1151795604700004352, 1179438699212804456448L, 1207745227993911763402752L, 1236731113465765645724418048L, 1266412660188944021221804081152L]
        self.colMasks = [1239150146850664126585242625L, 2478300293701328253170485250L, 4956600587402656506340970500L, 9913201174805313012681941000L, 19826402349610626025363882000L, 39652804699221252050727764000L, 79305609398442504101455528000L, 158611218796885008202911056000L, 317222437593770016405822112000L, 634444875187540032811644224000L]
        
        self.restart()
    
    def restart(self):
        board = 0 # encoded by 10x10 = 100 bits
        tile = random.randrange(self.numTiles)
        self.state = (board << 5) + tile # additional 5 bits for the tile
        self.score = 0
        self.times = [0.0 for i in range(10)]
    
    def predictNextState(self, action, state=None): # return one possible state, reward, afterstate, and validity of the action
        # t0 = time.time()
        if state == None: state = self.state
        row = action // self.size
        col = action % self.size
        # self.times[0] += time.time() - t0
        
        # get the tile
        # t1 = time.time()
        tile = state & 0b11111
        board = (state >> 5)
        tileWidth = self.tilesWidth[tile]
        tileHeight = self.tilesHeight[tile]
        # self.times[1] += time.time() - t1
        
        # t8 = time.time()
        if (row < 0) or (col < 0) or (row > self.size - tileHeight) or (col > self.size - tileWidth):
            return (state, 0, state, 0)
        
        tileOnBoard = (self.tilesMask[tile] << ((self.size-tileWidth-col)+(self.size-tileHeight-row)*10))
        if (board & tileOnBoard) != 0:
            return (state, 0, state, 0)
        
        # self.times[8] += time.time() - t8
        
        
        # put the tile on the board
        # t2 = time.time()
        nextBoard = board | tileOnBoard
        nextBoard2 = nextBoard # make a copy of the next board
        # self.times[2] += time.time() - t2
        
        # check the completed rows or cols only on the filled rows and cols
        # t3 = time.time()
        completedRowsCols = 0
        for r in self.rowMasks[self.size-(row+tileHeight):self.size-row]:
            if (nextBoard & r) == r:
                nextBoard2 ^= r
                completedRowsCols += 1
        # self.times[3] += time.time() - t3
        
        # t4 = time.time()
        for c in self.colMasks[self.size-(col+tileWidth):self.size-col]:
            if (nextBoard & c == c):
                nextBoard ^= c
                completedRowsCols += 1
        # self.times[4] += time.time() - t4
        
        # remove the completed rows and columns
        # t5 = time.time()
        nextBoard &= nextBoard2
        
        # calculate the reward
        reward = self.tilesSize[tile] + completedRowsCols * (completedRowsCols+1) * 5
        # self.times[5] += time.time() - t5
        
        # generate one the possible state
        # t6 = time.time()
        afterstate = (nextBoard << 5)
        # self.times[6] += time.time() - t6
        # t7 = time.time()
        nextState = afterstate + int(random.random()*self.numTiles)
        # self.times[7] += time.time() - t7
        
        return (nextState, reward, afterstate, 1)
    
    def getNextState(self, afterstate): #return one possible state, given an afterstate
        return afterstate + random.randrange(self.numTiles)
    
    def act(self, action):
        # generate a possible next state
        [nextState, reward, afterstate, valid] = self.predictNextState(action, self.state)
        
        # update the score and the state
        self.score += reward
        self.state = nextState
        
        return reward
        
    def isTerminated(self, state=None):
        if state == None: state = self.state
        
        tile = state & 0b11111
        board = (state >> 5)
        mask = self.tilesMask[tile]
        tileHeight = self.tilesHeight[tile]
        tileWidth = self.tilesWidth[tile]
        
        rowMask = (1 << (self.size * tileHeight)) - 1
        
        colArray = xrange(self.size-tileWidth+1)
        for row in xrange(self.size-tileHeight+1):
            
            board2 = int(board & rowMask)
            
            for col in colArray:
                if (board2 >> col) & mask == 0: return False
            
            board >>= self.size
        return True
    
    def getValidActions(self, state=None):
        if state == None: state = self.state
        
        validActions = []
        tile = state & 0b11111
        board = (state >> 5)
        mask = self.tilesMask[tile]
        tileWidth = self.tilesWidth[tile]
        tileHeight = self.tilesHeight[tile]
        
        # get the mask for the nrows (n = tileHeight)
        rowMask = (1 << (self.size * tileHeight)) - 1
        
        # save some variables for optimisation
        append = validActions.append
        colArray = xrange(self.size-tileWidth,-1,-1)
        
        # iterate for every possible rows and columns
        for row10 in xrange((self.size-tileHeight)*10,-1,-10):
            
            # get the nrows of the board
            board2 = int(board & rowMask)
            
            for col in colArray:
                if board2 & mask == 0: append(row10+col)
                board2 >>= 1
            
            board >>= self.size
        
        return validActions
    
    def render(self):
        board = self.state >> 5
        tile = self.state & 0b11111
        shift = self.size*self.size-1
        print('Score: %d' % self.score)
        
        # print the board
        s = ''
        for r in range(self.size):
            for c in range(self.size):
                s += '#' if (board >> shift) & 0b1 else '.'
                shift -= 1
            s += '\n'
        print(s)
        print('')
        
        # print the tile
        t = [['.' for i in range(self.size)] for j in range(5)]
        for i in range(len(self.tiles[tile])):
            for j in range(len(self.tiles[tile][i])):
                t[i][j] = '#' if self.tiles[tile][i][j] == 1 else '.'
        b = [''.join(tt) for tt in t]
        print('\n'.join(b))

# play the game
def main():
    game = Game1010()
    while not game.isTerminated():
        game.render()
        action = input("Your action: ")
        game.act(action)

# simulate and time
def simulate(N=100):
    game = Game1010()
    # times = [0 for i in range(4+len(game.times))]
    nsteps = 0
    
    # tt0 = time.time()
    for j in range(N):
        game.restart()
        while True:
            nsteps += 1
            # t0 = time.time()
            possibleActions = game.getValidActions()
            # t1 = time.time()
            action = possibleActions[random.randrange(len(possibleActions))]
            # t2 = time.time()
            game.act(action)
            # t3 = time.time()
            terminated = game.isTerminated()
            # t4 = time.time()
            
            # times[0] += t1 - t0
            # times[1] += t2 - t1
            # times[2] += t3 - t2
            # times[3] += t4 - t3
            if terminated: break
        
        # for i in range(len(game.times)): times[4+i] += game.times[i]
    # print("Finishes in %f s" % (time.time() - tt0))
    # print(times)
    # print(nsteps)

if __name__ == "__main__":
    simulate(10000)
    # main()

File: assets/codes/1010/game_bit_operation.cc
#include <vector>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <inttypes.h>
using namespace std;

#define SIZE 10
#define NUM_TILES 19
#define BOARD_SIZE (SIZE*SIZE)

// #define TIMED
#ifdef TIMED
    // #define TIMED_ACT
    // #define TIMED_VALID_ACTIONS
#endif

typedef struct {
    __int128 board;
    int tile;
} State;

typedef struct {
    State afterstate;
    int reward;
    bool valid;
} ActionOutput;

// profiling
#ifdef TIMED
const int NUM_TIMES = 20;
clock_t times[NUM_TIMES] = {0};
#endif

// tiles parameters
static const int tilesWidth[]  = {1, 2, 1, 3, 1, 2, 2, 2, 2, 1, 4, 2, 1, 5, 3, 3, 3, 3, 3};
static const int tilesHeight[] = {1, 1, 2, 1, 3, 2, 2, 2, 2, 4, 1, 2, 5, 1, 3, 3, 3, 3, 3};
static const int tilesSize[]   = {1, 2, 2, 3, 3, 3, 3, 3, 3, 4, 4, 4, 5, 5, 5, 5, 5, 5, 9};
static const int tilesSizeCumSum[] = {0, 1, 3,  5,  8, 11, 14, 17, 20, 23, 27, 31, 35, 40, 45, 50, 55, 60, 65}; // starting index of the tile
static const int tiles[] = {
    0,                              // 0: #
    0, 1,                           // 1: ##
    0, 10,                          // 2: #/#
    0, 1, 2,                        // 3: ###
    0, 10, 20,                      // 4: #/#/#
    1, 10, 11,                      // 5: .#/##
    0, 10, 11,                      // 6: #./##
    0, 1, 10,                       // 7: ##/#.
    0, 1, 11,                       // 8: ##/.#
    0, 10, 20, 30,                  // 9: #/#/#/#
    0, 1, 2, 3,                     // 10: ####
    0, 1, 10, 11,                   // 11: ##/##
    0, 10, 20, 30, 40,              // 12: #/#/#/#/#
    0, 1, 2, 3, 4,                  // 13: #####
    0, 1, 2, 12, 22,                // 14: ###/..#/..#
    0, 1, 2, 10, 20,                // 15: ###/#../#..
    0, 10, 20, 21, 22,              // 16: #../#../###
    2, 12, 20, 21, 22,              // 17: ..#/..#/###
    0, 1, 2, 10, 11, 12, 20, 21, 22 // 18: ###/###/###
};
static const long tilesInt[] = { // mask of the tiles
    1,
    3,
    1025,
    7,
    1049601,
    3074,
    3073,
    1027,
    2051,
    1074791425,
    15,
    3075,
    1100586419201,
    31,
    4198407,
    1049607,
    7341057,
    7344132,
    7347207
};

static const __int128 rowMask = (1 << SIZE) - 1;
static const __int128 colMask = ((__int128)(1239150)*1000000+146850)*1000000000000000+664126585242625;

// game class
class Game1010 {
  public:
    State state;
    int score;
    
    Game1010();
    
    // next state predictions
    ActionOutput getAfterState(int, State);
    State getNextState(State);
    vector<State> getAllStates(State);
    
    // action
    void act(int);
    vector<int> getValidActions2(State);
    vector<int> getValidActions(State);
    
    // termination and restart
    bool isTerminated(State);
    void restart();
    
    // user-interaction
    void render();
};


Game1010::Game1010() {
    restart();
}

// get afterstate for a given action and the state
ActionOutput Game1010::getAfterState(int action, State inputstate) {
    // determine the row and column of the action
    #ifdef TIMED_ACT
    clock_t t4 = clock();
    #endif
    int row = action / SIZE;
    int col = action % SIZE;
    
    // extract the tile information
    const int tileSize = tilesSize[inputstate.tile];
    const int tileHeight = tilesHeight[inputstate.tile];
    const int tileWidth = tilesWidth[inputstate.tile];
    const int tileInt = tilesInt[inputstate.tile];
    
    #ifdef TIMED_ACT
    times[4] += clock() - t4;
    
    clock_t t5 = clock();
    #endif
    
    // check if the action is valid
    bool isValid = true;
    if ((row < 0) || (col < 0) || (row > SIZE - tileHeight) || (col > SIZE - tileWidth)) { // check if the tiles are inside the box
        isValid = false;
    }
    // check if all tiles to be placed are blank
    else if ((long)(inputstate.board >> action) & tileInt) {
        isValid = false;
    }
    #ifdef TIMED_ACT
    times[5] += clock() - t5;
    #endif
    
    // if the action is not valid, return the next state as it is
    if (!isValid) {
        ActionOutput ao;
        ao.afterstate = inputstate;
        ao.reward = 0;
        ao.valid = false;
        return ao;
    }
    
    #ifdef TIMED_ACT
    clock_t t6 = clock();
    #endif
    // obtain the tile mask to be placed on board
    __int128 tileOnBoard = (__int128)(tileInt) << action;
    
    // if it's valid, then place the tiles on the board
    State nextstate = inputstate;
    nextstate.board = inputstate.board | tileOnBoard;
    
    // copy the board
    __int128 boardRow = nextstate.board;
    __int128 boardCol = nextstate.board;
    
    #ifdef TIMED_ACT
    times[6] += clock() - t6;
    
    clock_t t7 = clock();
    #endif
    // check the completed rows and columns, then remove them
    int ncomplete = 0;
    for (int i = row*SIZE; i < (row+tileHeight)*SIZE; i+=SIZE) {
        if (((nextstate.board >> i) & rowMask) == rowMask) {
            ++ncomplete;
            boardRow ^= (rowMask << i);
        }
    }
    for (int j = col; j < (col+tileWidth); ++j) {
        if (((nextstate.board >> j) & colMask) == colMask) {
            ++ncomplete;
            boardCol ^= (colMask << j);
        }
    }
    #ifdef TIMED_ACT
    times[7] += clock() - t7;
    
    clock_t t8 = clock();
    #endif
    
    // copy back to the nextstate board
    nextstate.board = boardRow & boardCol;
    
    #ifdef TIMED_ACT
    times[8] += clock() - t8;
    
    clock_t t10 = clock();
    times[10] += clock() - t10;
    #endif
    
    // calculate the reward
    int reward = tileSize + (ncomplete+1)*ncomplete*5;
    
    // return the output_iterator
    ActionOutput ao;
    ao.afterstate = nextstate;
    ao.reward = reward;
    ao.valid = true;
    return ao;
}

// get one possible next state after the afterstate
State Game1010::getNextState(State afterstate) {
    State ss = afterstate;
    ss.tile = rand() % NUM_TILES;
    return ss;
}

vector<State> Game1010::getAllStates(State afterstate) {
    vector<State> allStates;
    allStates.resize(NUM_TILES);
    
    // the board for the next state is the same with afterstate, but the tile is different
    for (int i = 0; i < NUM_TILES; ++i) {
        State ss = afterstate;
        ss.tile = i;
        allStates[i] = ss;
    }
    
    return allStates;
}

// do an action to the board
void Game1010::act(int action) {
    // get the next state
    ActionOutput ao = getAfterState(action, state);
    if (ao.valid) {
        state = getNextState(ao.afterstate);
        
        // add the score with the reward
        score += ao.reward;
    }
}

// list all possible valid actions
vector<int> Game1010::getValidActions(State inputstate) {
    // extract the tile information
    #ifdef TIMED_VALID_ACTIONS
    clock_t t11 = clock();
    #endif
    
    int hspaceSIZE = (SIZE-tilesHeight[inputstate.tile])*SIZE;
    int wspace = SIZE-tilesWidth[inputstate.tile];
    
    #ifdef TIMED_VALID_ACTIONS
    times[11] += clock() - t11;
    
    clock_t t12 = clock();
    #endif
    // prepare the output vector
    vector<int> validActions;
    validActions.resize(BOARD_SIZE);
    int nActions = 0;
    
    #ifdef TIMED_VALID_ACTIONS
    times[12] += clock() - t12;
    #endif
    
    // check for every possible actions
    long boardInt64;
    long tileInt64 = tilesInt[inputstate.tile];
    for (int row = 0; row <= hspaceSIZE; row+=SIZE) {
        
        // cast the lsb 64 bits to long integer to speed up
        boardInt64 = static_cast<long>(inputstate.board >> row);
        for (int col = 0; col <= wspace; ++col) {
            
            // if the tile positions are all zeros, then cast it as a valid action
            if (!(boardInt64 & tileInt64)) validActions[nActions++] = row + col;
            boardInt64 >>= 1;
        }
    }
    
    validActions.resize(nActions);
    return validActions;
}

// check if it's terminated (if there's no possible action)
bool Game1010::isTerminated(State inputstate) {
    // extract the tile information
    int tileWidth = tilesWidth[inputstate.tile];
    long tileInt = tilesInt[inputstate.tile];
    
    int hspace = SIZE-tilesHeight[inputstate.tile];
    int wspace = SIZE-tileWidth;
    
    // check all possible actions
    __int128 boardInt128 = inputstate.board;
    unsigned long boardInt64;
    for (int row = 0; row <= hspace; ++row) {
        
        // cast the lsb 64 bits to long integer to speed up
        boardInt64 = static_cast<unsigned long>(boardInt128);
        for (int col = 0; col <= wspace; ++col) {
            
            // for a certain action, check if the tiles are blank
            if (!(boardInt64 & tileInt)) return false;
            boardInt64 >>= 1;
        }
        boardInt128 >>= SIZE;
    }
    return true;
}

// restart the game: clear the board, assign a random tile, and reset the score
void Game1010::restart() {
    state.board = 0;
    state.tile = rand() % NUM_TILES; // generate random number between 0 and NUM_TILES-1
    
    // reset the score
    score = 0;
}

void Game1010::render() {
    // print the score
    printf("Score: %d\n", score);
    
    // print the board
    __int128 board = state.board;
    for (int i = 0; i < SIZE; ++i) {
        for (int j = 0; j < SIZE; ++j) {
            if (board & 1) printf("#");
            else printf(".");
            board >>= 1;
        }
        printf("\n");
    }
    printf("\n");
    
    // print the tile
    const int* tile = &tiles[tilesSizeCumSum[state.tile]];
    int tileSize = tilesSize[state.tile];
    int nTile = 0;
    for (int i = 0; i < 5; ++i) {
        for (int j = 0; j < 5; ++j) {
            if ((nTile < tileSize) && (tile[nTile] == i*SIZE+j)) {
                printf("#");
                ++nTile;
            }
            else printf(".");
        }
        printf("\n");
    }
    
}


void play(bool random=false, bool wait=true, bool render=true) {
    int action;
    Game1010* g = new Game1010();
    if (render) g->render();
    while (!g->isTerminated(g->state)) {
        if (wait)
            scanf("%i", &action);
        if (random) {
            vector<int> validActions = g->getValidActions(g->state);
            action = validActions[rand() % validActions.size()];
        }
        if (render) printf("Your action: %d\n", action);
        g->act(action%BOARD_SIZE);
        if (render) g->render();
    }
    delete g;
}

void simulate(int N) {
    
    int action;
    Game1010* g = new Game1010();
    int nsteps = 0;
    bool isTerminated;
    
    for (int i = 0; i < N; ++i) {
        while (true) {
            #ifdef TIMED
            clock_t t0 = clock();
            #endif
            vector<int> validActions = g->getValidActions(g->state);
            #ifdef TIMED
            clock_t t1 = clock();
            #endif
            action = validActions[rand() % validActions.size()];
            #ifdef TIMED
            clock_t t2 = clock();
            #endif
            g->act(action);
            #ifdef TIMED
            clock_t t3 = clock();
            #endif
            isTerminated = g->isTerminated(g->state);
            #ifdef TIMED
            clock_t t4 = clock();
            #endif
            
            #ifdef TIMED
            times[0] += t1 - t0;
            times[1] += t2 - t1;
            times[2] += t3 - t2;
            times[3] += t4 - t3;
            #endif
            
            ++nsteps;
            if (isTerminated) break;
        }
        g->restart();
    }
    printf("Total steps: %d\n", nsteps);
    #ifdef TIMED
    for (int i = 0; i < NUM_TIMES; ++i) {
        printf("%f, ", (float)times[i] / CLOCKS_PER_SEC);
    }
    #endif
    printf("\n");
    delete g;
}

int main() {
    // play(false, true, true);
    simulate(100000);
}































